core/domain/stats_services.py

```diff
@@ -1024,3 +1025,198 @@ def _get_calc_output(exploration_id, state_name, calculation_id):
             calculation_output)
     else:
         return None
+
+
+def change_learner_answer_details(
+        state_reference, entity_type, interaction_id, change):
+    """Changes the learner answer details object, as per the changes received
+    and then returns it.
+
+    Args:
+        state_reference: str. This is used to refer to a state
+            in an exploration or question. Like for an exploration the
+            value will be equal to 'exp_id.state_name' & for question
+            this will be equal to 'question_id' only.
+        entity_type: str. The type of entity i.e 'exploration' or 'question '.
+        interaction_id: str. The ID of the interaction.
+        change: LearnerAnswerDetailsChange. The change to be applied
+            to the LearnerAnswerDetails object.
+
+    Returns:
+        LearnerAnswerDetails. The resulting learner answer details object,
+            after applying the change.
+     """
+    learner_answer_details = get_learner_answer_details(
+        state_reference, entity_type)
+    if change.cmd == stats_domain.CMD_RECORD_LEARNER_ANSWER_INFO:
+        if not learner_answer_details:
+            learner_answer_details = stats_domain.LearnerAnswerDetails(
+                state_reference, entity_type, interaction_id, [], 0)
+        learner_answer_info_id = (
+            stats_domain.LearnerAnswerInfo.get_learner_answer_info_id(
+                state_reference, entity_type))
+        learner_answer_info = stats_domain.LearnerAnswerInfo(
+            learner_answer_info_id, change.answer,
+            change.answer_details, datetime.datetime.utcnow())
+        learner_answer_details.add_learner_answer_info(learner_answer_info)
+    elif change.cmd == stats_domain.CMD_DELETE_LEARNER_ANSWER_INFO:
```

Comment:
Here, if learner_answer_details is None, throw InvalidInputError

Reply:
Done